### PR TITLE
protobuf3: 21.12 -> 23.4

### PR DESCRIPTION
**Pull request by tobim**
###### Description of changes

https://github.com/protocolbuffers/protobuf/releases/tag/v23.4

v22 is also not packaged, but since it hasn't been added so far I think we can leave a gap.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/abseil-cpp/202301.nix [hunk 1]
```diff
@@ -18,13 +19,17 @@ stdenv.mkDerivation (finalAttrs: {
   };
 
   cmakeFlags = [
+    "-DABSL_BUILD_TEST_HELPERS=ON"
+    "-DABSL_USE_EXTERNAL_GOOGLETEST=ON"
```
Comment (marsam) [inline pkgs/development/libraries/abseil-cpp/202301.nix:23]: Is this needed for https://github.com/abseil/abseil-cpp/issues/1407? if so please add a comment about it

File: pkgs/development/libraries/abseil-cpp/202301.nix [hunk 2]
```diff
@@ -18,13 +19,17 @@ stdenv.mkDerivation (finalAttrs: {
...
     "-DBUILD_SHARED_LIBS=${if static then "OFF" else "ON"}"
   ] ++ lib.optionals (cxxStandard != null) [
     "-DCMAKE_CXX_STANDARD=${cxxStandard}"
   ];
 
   nativeBuildInputs = [ cmake ];
 
+  buildInputs = [ gtest ];
```
Comment (marsam) [inline pkgs/development/libraries/abseil-cpp/202301.nix:31]: just curious, can gtest be in `nativeBuildInputs`?

File: pkgs/development/libraries/abseil-cpp/202301.nix (hunk 1 above)
Comment (tobim) [reply pkgs/development/libraries/abseil-cpp/202301.nix:23]: That is unrelated, abseil just won't look for the provided gtest if this option is not set.

File: pkgs/development/libraries/abseil-cpp/202301.nix (hunk 2 above)
Comment (tobim) [reply pkgs/development/libraries/abseil-cpp/202301.nix:33]: I double checked with `strictDeps` and it cannot be native.

Comment (tobim) [reply pkgs/development/libraries/abseil-cpp/202301.nix:33]: I updated the PR to set `strictDeps = true`.

**Review by marsam — APPROVED** [review]

**Comment by tjni** [comment]
protobuf-c is broken and it looks like they haven't released a version that supports 22.x or 23.x yet. I did find https://github.com/protobuf-c/protobuf-c/pull/673, which indicates support should be present in their `next` branch. 

@NickCao, do you have a preference on what to do?

**Comment by NickCao** [comment]
> @NickCao, do you have a preference on what to do?

Let's bump to the version in `next`, `protobufc` is depended by a handful of important packages and we cannot leave it broken.

**Comment by tjni** [comment]
I opened https://github.com/NixOS/nixpkgs/pull/248306. Adding abseil-cpp to propagatedBuildInputs was indeed needed. I don't have the ability to test this on Linux; I could only test it on aarch64-darwin since I had already built all of the stuff on staging.

**Pull request merged**